sematic/abstract_future.py

```diff
@@ -382,3 +382,25 @@ def get_next_timeout(futures: List[AbstractFuture]) -> TimeoutFuturePair:
         return None, None
 
     return min(remaining_time_future_pairs)
+
+
+def clone(future: AbstractFuture) -> AbstractFuture:
```

Comment:
> or make it an instance method of `AbstractFuture`.

No, because we want to preserve the option that we support things like: `my_func().foo()` at a later point, and we don't want to pollute the `AbstractFuture` namespace.

> Also, we have a bunch of cloning logic in factories.py. Shall we consolidate?

No, because (a) that creates a circular dependency of factories -> future -> resolver (b) the cloning depends on implementation details of `AbstractFuture` (knowledge of which attributes are set post-init) and thus should be co-located.